docs/proposals/20190709-cluster-spec-crds.md

```diff
@@ -0,0 +1,325 @@
+
+The Cluster controller does not handle these fields directly. Instead, the provider-specific actuator receives the `ProviderSpec` as part of the Cluster object and sets the `ProviderStatus` to reflect the status of the cluster infrastructure. 
+
+This proposal outlines the replacement of these provider-specific embedded fields by references to objects managed by a provider controller, effectively eliminating the actuator interface. The new objects are introduced and the cooperation between the Cluster controller and the Cluster Infrastructure Controller is outlined. This process introduces a new field `InfrastructureState` which reflects the state of the provisioning process.
+
+## Motivation
+
+Embedding opaque provider-specific information has some disadvantages: 
+- Using an embedded provider spec as a `RawExtension` means its content is not validated as part of the Cluster object validation
+- Embedding the provider spec makes difficult implementing the logic of the provider as an independent controller watching for changes in the specs. Instead, providers implement cluster controllers which embed the generic logic and a provider-specific actuator which implements the logic for handling events in the Cluster object.  
+- Embedding the provider status as part of the Cluster requires that either the generic cluster controller be responsible for pulling this state from the provider (e.g. by invoking a provider actuator) or the provider updating this field, incurring in overlapping responsibilities. 
+
+### Goals
+
+1. To use Kubernetes Controllers to manage the lifecycle of provider specific cluster infrastructure object.
+1. To validate provider-specific object as early as possible (create, update).
+1. Allow cluster providers to expose the status via a status subresource, eliminating the need of status updates across controllers.
+
+### Non-Goals/Future Work
+1. To modify the generic Cluster object beyond the elements related to the provider-specific specs and status.
+1. Splitting the provider-specific infrastructure into more granular elements, such as network infrastructure and control plane specs. Revisiting the definition of the provider specific infrastructure is left for further proposals.
+1. To replace the cluster status by alternative representations, such as a list of conditions.
+1. To propose cluster state lifecycle hooks. This must be part of a future proposal that builds on these proposed changes
+
+## Proposal
+
+This proposal introduces changes in the Cluster data model and describes a collaboration model between the Cluster controller and the provider infrastructure controller, including a model for managing the state transition during the provisioning process. 
+
+### Data Model changes
+
+```go
+type ProviderSpec struct
```

Comment:
Are you suggesting we keep `ProviderSpec`, so that in a `Cluster` we'd have `.Spec.ProviderSpec.InfrastructureRef`? I was hoping this would match the v1alpha2 changes for `Machine`, where we removed `ProviderSpec` entirely, replacing it with the `Bootstrap` and `InfrastructureRef` fields.

Reply:
+1 This can definitely be top-level